File: lecture/dingcorithm/2st_week/02_13_get_kth_node_from_last.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self, value):
        self.head = Node(value)

    def append(self, value):
        cur = self.head
        while cur.next is not None:
            cur = cur.next
        cur.next = Node(value)

    def get_kth_node_from_last(self, k):
        # 구현해보세요!
        # 끝까지 돌아간다. 몇개인지 파악해야?
        cur = self.head
        next = cur.next
        last_index = 1

        while next is not None:
            cur = next
            last_index +=1
            next = cur.next


        print('last index',last_index)
        # 뒤에서 k만큼 되돌아간다.

        result = self.head
        for i in range(last_index-k):
            result = result.next

        return result


linked_list = LinkedList(6)
linked_list.append(7)
linked_list.append(8)
linked_list.append(9)

print(linked_list.get_kth_node_from_last(2).data)  # 8이 나와야 합니다!

File: study/dingcorithm/2st_week/02_13_get_kth_node_from_last.js
// 🔁 복습 필요 (3/26)(3/30)

class LinkedListKthFromLast {
  constructor(value) {
    this.head = new LinkedListKthFromLast.Node(value);
  }

  append(value) {
    let cur = this.head;
    while (cur.next !== null) {
      cur = cur.next;
    }
    cur.next = new LinkedListKthFromLast.Node(value);
  }

  getKthNodeFromLast(k) {
    let cur = this.head;
    let cur_index = 1;
    let whole_len = 0;

    // 전체 길이를 구하기
    // k번째
    // 15323
    // 전체 길이 5 끝에서 2 => 2 앞에서 4번째 5-2+1 = 4
    // 전체 길이 5 끝에서 4 => 5 앞에서 2번째 5-4+1 = 2
    // 그런데 현재 전체 길이를 모름

    while (cur !== null) {
      cur = cur.next;
      whole_len += 1;
    }

    cur = this.head;

    while (cur !== null) {
      if (cur_index === whole_len - k + 1) {
        return cur;
      }
      cur_index += 1;
      cur = cur.next;
    }
  }
}

// Define Node as a static property of LinkedListKthFromLast
LinkedListKthFromLast.Node = class {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
};

// Main execution
const linkedList = new LinkedListKthFromLast(6);
linkedList.append(7);
linkedList.append(8);
linkedList.append(4);
linkedList.append(2);

console.log(linkedList.getKthNodeFromLast(2).data); // 8이 나와야 합니다!
